### DISCUSSION TITLE
2nd probe pin ?

### DISCUSSION THREAD

**Discussion by Tpacnc** [fra589/grbl-Mega-5X#276] [Q&A] [answered]
Hi just Wanting to know if a 2nd probe pin can be configured for use with a tool setter and 3D Touch probe.
My problem is that my 3D Touch probe operates in NO (normally open) and my tool setter operates in NC (normally closed) so I can't just have them both go to the probe pin.
So I was thinking maybe I could use one of the spare pins for the tool setter that I could invert like is possible with the standard probe pin and have grbl use it like as if I just had 2 NO(normally open) switches going to the same probe pin.
But I have no idea how I would go about configuring a spare pin to do that or if it's even possible.

I figured out a work around for now using a relay to switch the NC(normally closed)state of my tool setter to NO(normally open)meaning they can then be connected to the same probe pin however I don't think it is the best way of doing it for repeatability
due to the time it takes to switch the relay and I am concerned it may affect the measurement of the tool.  

![2nd probe pin work around](https://user-images.githubusercontent.com/108869389/177864848-06f1debe-5d56-4ab6-897f-2e806417fe52.jpg)

**Comment by fra589** [comment, marked as answer]
Hi @Tpacnc,  
  
Using a relay is a good id, but not like you think :)
You can use a relay to drive the choice of which probe tool is connected, using the digital output functionality to drive the relay:
![Probe-NC-NO](https://user-images.githubusercontent.com/19529154/177940806-baa18561-4959-420d-bf27-47866a6769dc.svg)
If you use D16 which is the default 1st digital output number 0, you may use the GCode command M62/63/64 or 65 P0 to drive which probe tool you are using.

- M62 P- - turn on digital output synchronized with motion. The P- word specifies the digital output number.
- M63 P- - turn off digital output synchronized with motion. The P- word specifies the digital output number.
- M64 P- - turn on digital output immediately. The P- word specifies the digital output number.
- M65 P- - turn off digital output immediately. The P- word specifies the digital output number.
  
See http://linuxcnc.org/docs/html/gcode/m-code.html#mcode:m62-m65 for details.
  
If you connect your probes like the drawing above, when the digital output will be down, the tool setter will be active, and when the digital output will be up, the 3D touch will be active.
  
@++;  
Gauthier.

**Reply by Tpacnc** [reply]
Hi 
Thank you for the response.
so I have temporarily set up the tool setter and 3D touch probe on my machine last night using the circuit I originally posted. It seems to work pretty well as far as I can see.
The 3D touch probe works as it should to find the corner of a work piece etc and the the tool setter works to measure the Hight of my end mill and doesn't look to have too much delay between the end mill touching the tool setter to the machine stopping and retracting using the relay.
However I like your idea of using the M62-M65 to control which tool is active (with only one active at a time less chance of one being knocked and getting triggered messing something up) and I think I will try it this way but I think I would still have to use a 2nd relay to switch the tool setters resting state(NC when no end mill is pressing it) to NO

![M62-M65 tool setter and probe](https://user-images.githubusercontent.com/108869389/178047425-d8553690-0fe9-44a6-8cdd-8e01421b01a6.jpg)

Otherwise as soon as the M62-M65 lines are sent to switch from the 3D touch probe to the tool setter it would trigger the pin due to the mismatch of a NO 3D touch probe and a NC tool setter I believe.

To explain my thoughts behind relay 1 in the diagram a bit more:
While ever the Arduino is powered relay 1 is powered and triggered keeping COM connected to NO, turning of the relay using the tool setter to disconnect the purple wire from GND, would then connect COM to NC (converting the NC resting state of the tool setter to NO), connecting the probe pin to ground(assuming the M62-M65 command is triggering relay 2)

**Reply by fra589** [reply]
Hi @Tpacnc,  
  
> Otherwise as soon as the M62-M65 lines are sent to switch from the 3D touch probe to the tool setter it would trigger the pin due to the mismatch of a NO 3D touch probe and a NC tool setter I believe.
  
You have 2 solutions for this:  
1. Issue a `$6=<0/1>` (Probe pin invert) parameter change command to inverts the probe input pin signal just after the M62-M65 command,
2. Leaving always `$6=0`, Use `G38.2` or `G38.3` for the NO probe and `G38.4` or `G38.5` for NC probe.  
cf. http://linuxcnc.org/docs/stable/html/gcode/g-code.html#gcode:g38

@++;
Gauthier.

**Reply by Tpacnc** [reply]
Yep I have now set up both my probes as per your wiring diagram using the M62-M65 to switch between probes and have added M64 P0 $6=1 to the start of my semiauto tool change macro and M63 P0 $6=0 to the end.
It now seems to work flawlessly, didn't know realise you could change the $ settings on the fly like that as part of a macro.

Thank you for the help very much appreciated

**Reply by langwadt** [reply]
settings are in eeprom so repeatedly changing them might not be the best idea

**Reply by Tpacnc** [reply]
update 
so after seeing your reply @langwadt I have got it working using @fra589 2nd method by getting rid of the $6=1 / $6=0 at the start and end of the macro keeping the M64 P0 at the start, M63 P0 at the end and replacing all of the G38.2 commands with G38.4
this works just as good as @fra589 1st solution. and thinking about it this would be the preferred method seeing as non of the $ setting are changed and in the event of something going wrong, like the machine freezing up mid way through the macro before getting to the $6=0 to change the pin state back, you would have to remember to switch it back manually after resetting.

Thanks again @fra589 and thank you @langwadt for steering me towards the better solution
